### ISSUE TITLE
Provide a callback to decrypt DB passwords

### ISSUE THREAD

**Issue by twicksell** [spring-projects/spring-boot#1312]
Common requirement is to encrypt your db password and put it in your props file. It'd be nice for Boot to provide some kind of callback so that I can provide decryption logic. Perhaps consider making it flexible so we can use it for more than just the DB passwords. My initial thought was to register a custom spel function and create a property like
spring.datasource=#{decrypt('someCrazyEncryptedPassword')}

**Comment by hohwille** [comment]
Just to answer some of the concerns why this could be helpful. In general this is some sort of security by obscurity if in the end the application can decrypt the passwords. However, the question is more how to do a safe packaging and distribution of an application.
With spring boot you can easily externalize the configurations (./config/application.properties) and build portable deliverables. However, if you want automated deployments of the configurations per environment you might want to build config packages as RPM, deb, yum, etc.
With the suggested approach you could have the production settings in maven filters in a VCS where every developer can access but as the passwords are strongly encrypted the passwords are still secure.
Then only e.g. a keystore or some additional secret is required on the machine for the installation beside the packages build and deployed (via DevOps). So IMHO this is a nice feature.
As it seems Jasypt already solved this with spring boot but if there was official support designed by the spring makers this would be even better as I also fear the chicken-egg-problem when things get more complex.

**Comment by dsyer** [comment]
As I said above, I don't think Jasypt cuts it for me. Please take a look at the features in Spring Cloud Context (moved from Spring Cloud Config during the 1.0.x releases). All the encryption support is in a single, small jar with only Spring Boot dependencies. It would be dumb to duplicate that here.

**Comment by hohwille** [comment]
Fine. I will have a look. Thanks for the hint.

Could you also have a look at this and consider some rethinking about springs inner design:
https://github.com/ulisesbocchio/jasypt-spring-boot/issues/7

I am pretty much aware that spring is a swiss army knife and people will use it in million different (and sometimes sick) ways. So regression is not an easy one, but I would be happy, if you could have a look and see if there is something wrong in spring that can be fixed. Thanks!

**Comment by dsyer** [comment]
@ulisesbocchio explains in that issue that it is *his* design decision not to cache decrypted values. I don't see what it has to do with Spring Boot yet. Please let us know separately if you find something that needs changing.

**Comment by hohwille** [comment]
I filed a separate issue for this so I am done here.

**Comment by hohwille** [comment]
> Please take a look at the features in Spring Cloud Context

So you are talking about things like this:
https://github.com/spring-cloud/spring-cloud-commons/blob/master/spring-cloud-context/src/main/java/org/springframework/cloud/context/encrypt/EncryptorFactory.java

Sorry to say so but this is rather disappointing. Actually it would be better to allow to configure a custom decrytpor bean that is used for property values in the ENC(...) form. Then your current hardwired implementation can be a fallback if nothing else is configured.

**Comment by mminella** [comment]
For the record, there is a related Spring Core issue for similar functionality that's been around for a while: https://jira.spring.io/browse/SPR-10666

**Comment by arafique** [comment]
Hi Guys

Can anyone please help me with following issue

http://stackoverflow.com/questions/31824601/spring-boot-not-loading-propertysourceloader

Thanks

**Comment by ulisesbocchio** [comment]
@arafique For what I can tell the `SpringFactoriesLoader` mechanism for factory `PropertySourceLoader` is only used by `PropertySourcesLoader` which in turn is ONLY used to load the application properties. Those are either `application.properties`, `application.yml`, or `application.yaml`. So for your example just add a file `application.properties` to your classpath and you'll get the exception you are expecting.
What I don't know is why other property source import mechanisms such as using annotation @PropertySource are not going through `PropertySourcesLoader` to take advantage of the factories loaders and override mechanism.

**Comment by DanieleTorino** [comment]
@stephane-deraco: How did you get Spring to use your `PropertySourceLoader`? I tried your approach, but I can't seem to get Spring Boot to use my component.

**Comment by stephane-deraco** [comment]
@DanieleTorino In fact, I have followed the advice given by @dsyer and took a look at Spring cloud client. However, I do not use Spring cloud.

I came to this solution inspired by https://github.com/spring-cloud/spring-cloud-commons/blob/cde7c7f3118382490c28776f66e0a56f248141fd/spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializer.java :

Create a class implementing `ApplicationContextInitializer<ConfigurableApplicationContext>` and annotated with `@Configuration`.

Initialize your encryptor with something like :
```java
// Try to initialize the encryptor with master password
encryptor = new StandardPBEStringEncryptor();
final char[] password = // Get your pass here
((StandardPBEStringEncryptor) encryptor).setPasswordCharArray(password);
Arrays.fill(password, '\0');
```

And then write:

```java
	@Override
	public void initialize(ConfigurableApplicationContext applicationContext) {
		ConfigurableEnvironment environment = applicationContext.getEnvironment();
		Map<String, Object> overrides = new LinkedHashMap<>();
		for (PropertySource<?> source : environment.getPropertySources()) {
			decrypt(source, overrides);
		}
		if (!overrides.isEmpty()) {
			environment.getPropertySources().addFirst(
					new MapPropertySource("decrypted", overrides));
		}
	}

	private void decrypt(PropertySource<?> source, Map<String, Object> overrides) {
		if (source instanceof EnumerablePropertySource) {
			EnumerablePropertySource<?> enumerable = (EnumerablePropertySource<?>) source;
			for (String key : enumerable.getPropertyNames()) {
				String value = source.getProperty(key).toString();
				if (value.startsWith("{cipher}")) {
					value = value.substring("{cipher}".length());
					try {
						value = encryptor.decrypt(value);
						LOG.debug("Decrypted: key={}", key);
					} catch (Exception e) {
						LOG.warn(String.format("Cannot decrypt: key=%s", key), e);
						// Set value to empty to avoid making a password out of the
						// cipher text
						value = "";
					}
					overrides.put(key, value);
				}
			}
		} else if (source instanceof CompositePropertySource) {
			for (PropertySource<?> nested : ((CompositePropertySource) source)
					.getPropertySources()) {
				decrypt(nested, overrides);
			}
		}
	}
```

This is something that works for me, including on Yaml properties file.

**Comment by DanieleTorino** [comment]
@stephane-deraco: Thanks a lot. I got it working with your help, my complete solution can be found on [stackoverflow](http://stackoverflow.com/questions/31989883/process-spring-boot-externalized-property-values).

**Comment by stephane-deraco** [comment]
@DanieleTorino You're welcome

**Comment by philwebb** [comment]
I'm going to close this because I think https://cloud.spring.io/spring-cloud-vault/ is the best way to solve this.

**Comment by PAX523** [comment]
> I came to this solution inspired by https://github.com/spring-cloud/spring-cloud-commons/blob/cde7c7f3118382490c28776f66e0a56f248141fd/spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializer.java :
> 
> Create a class implementing ApplicationContextInitializer<ConfigurableApplicationContext> and annotated with @Configuration.

...and if ``@Configuration`` doesn't do the trick then register your ``ApplicationContextInitializer`` before starting your Spring Boot application in main method via:

```
  public static void main(final String[] args) {
    final SpringApplication springApplication = new SpringApplication(MySpringBootService.class);
    springApplication.addInitializers(new ApplicationPropertiesDecryptor());
    springApplication.run(args);
  }
```
